src/rpc/mining.cpp

```diff
@@ -720,23 +720,34 @@ UniValue mkblocktemplate(const UniValue &params,
         CBlockIndex *pindexPrevNew = chainActive.Tip();
         nStart = GetTime();
 
-        // Create new block
-        if (!coinbaseScript)
+        // If client code didn't specify a coinbase address for the mining reward, grab one from the wallet.
+        if (coinbaseScript.empty())
         {
-            // if they didn't pass in a coinbaseScript, grab an address from wallet
-            GetMainSignals().ScriptForMining(coinbaseScript);
+            // Note that we don't cache the exact script from this to the prevCoinbaseScript -- it's sufficient
+            // to cache the fact that client code didn't specify a coinbase address (by caching the empty script).
+            boost::shared_ptr<CReserveScript> tmpScriptPtr;
+            GetMainSignals().ScriptForMining(tmpScriptPtr);
+
+            // throw an error if shared_ptr is not valid -- this means no wallet support was compiled-in
+            if (!tmpScriptPtr)
+                throw JSONRPCError(RPC_INTERNAL_ERROR,
+                    "Wallet support is not compiled-in, please specify an address for the coinbase tx");
+
+            // If the keypool is exhausted, the shared_ptr is valid but no actual script is generated; catch this.
+            if (tmpScriptPtr->reserveScript.empty())
+                throw JSONRPCError(
+                    RPC_WALLET_KEYPOOL_RAN_OUT, "Error: Keypool ran out, please call keypoolrefill first");
+
+            // Everything checks out, proceed with the wallet-generated address. Note that we don't tell the wallet to
+            // "KeepKey" this address -- which means future calls will return the same address from the wallet for
```

Comment:
Ok, seems like this is an almost never used case since people don't rely on wallets for both spending and mining AFAIK.  Its low priority so we'll leave to another PR if someone is inspired to do it.

Any block candidate provided using this API has to be submitted via submitminingsolution so there is a place.  But we can't call KeepKey at that point.  The user might have made another transaction, reusing the key, and then a block is found.

Reply:
> people don't rely on wallets for both spending and mining AFAIK.

Ah, from what I gathered too it seems that's the case. 

> But we can't call KeepKey at that point. The user might have made another transaction, reusing the key, and then a block is found.

I'm not super familiar with the wallet code but it seemed at first perusal that calling KeepKey on a key not in the keypool is what amounts to a no-op...?

Hey Andrew you're awesome man. Thanks for merging.

You did pique my interest about this use case.  I'll think about if I can tackle it in an awesome way at some point .. thanks.